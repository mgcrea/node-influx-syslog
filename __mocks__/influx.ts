// @NOTE https://jestjs.io/docs/en/manual-mocks.html

const d = (obj: any) => console.dir(obj, {depth: 10, colors: true});

const influx = jest.genMockFromModule('influx');

const __mockedWritePoints = jest.fn().mockImplementation(() => {
  d('writePoints');
});

export class InfluxDB {
  addSchema() {}
  async writePoints() {
    __mockedWritePoints();
  }
}

influx.InfluxDB = InfluxDB;
influx.__mockedWritePoints = __mockedWritePoints;

module.exports = influx;
