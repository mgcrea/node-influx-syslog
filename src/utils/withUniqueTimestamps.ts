import {IPoint} from 'influx';
import padStart from 'lodash.padstart';

const withUniqueTimestamps = (points: IPoint[]): IPoint[] => {
  const initialValue: {
    uniquePoints: IPoint[];
    lastTimestamp: number;
    currentNanoSecond: number;
  } = {uniquePoints: [], lastTimestamp: 0, currentNanoSecond: 0};

  const {uniquePoints} = points.reduce((soFar, value) => {
    // Only process fields with timestamps
    if (!value.fields || !value.fields.timestamp) {
      return soFar;
    }
    const {timestamp} = value.fields;
    // Order messages by "virtual" nano-seconds
    if (timestamp === soFar.lastTimestamp) {
      soFar.currentNanoSecond += 1;
    } else {
      soFar.currentNanoSecond = parseInt(timestamp.slice(-6), 10);
    }
    value.timestamp = `${timestamp.slice(0, -6)}${padStart(soFar.currentNanoSecond.toString(), 6, '0')}`;
    soFar.uniquePoints.push(value);
    soFar.lastTimestamp = timestamp;
    return soFar;
  }, initialValue);

  return uniquePoints;
};

export default withUniqueTimestamps;
