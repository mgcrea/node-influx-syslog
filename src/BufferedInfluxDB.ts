import {InfluxDB, IPoint, IWriteOptions} from 'influx';
import debounce from 'lodash.debounce';

import withUniqueTimestamps from 'src/utils/withUniqueTimestamps';

type BufferedInfluxDBOptions = {wait?: number; maxWait?: number};

class BufferedInfluxDB extends InfluxDB {
  private __stackedPoints: IPoint[];
  constructor(options: any, {wait = 1000}: BufferedInfluxDBOptions = {}) {
    super(options);
    this.__stackedPoints = [];
    this.__debouncedFlushPoints = debounce(this.__flushPoints, wait, {maxWait: wait});
  }
  async writePoints(points: IPoint[], options?: IWriteOptions): Promise<void> {
    this.__stackedPoints.push(...points);
    this.__debouncedFlushPoints(options);
  }
  async __flushPoints(options?: IWriteOptions): Promise<void> {
    const stackedPointsSoFar = this.__stackedPoints.splice(0, this.__stackedPoints.length);
    const uniqueStackedPoints = withUniqueTimestamps(stackedPointsSoFar);
    try {
      return await super.writePoints(uniqueStackedPoints, options);
    } catch (err) {
      // Restore stack on error
      this.__stackedPoints.unshift(...stackedPointsSoFar);
      throw err;
    }
  }
  // typings
  async __writePoints(_points: IPoint[], _options?: IWriteOptions): Promise<void> {}
  async __debouncedFlushPoints(_options?: IWriteOptions): Promise<void> {}
}

export default BufferedInfluxDB;
