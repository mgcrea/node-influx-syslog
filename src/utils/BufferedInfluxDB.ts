import {InfluxDB, IPoint, IWriteOptions} from 'influx';
import debounce from 'lodash.debounce';

import withUniqueTimestamps from 'src/utils/withUniqueTimestamps';

type BufferedInfluxDBOptions = {
  debounceWait?: number;
};

class BufferedInfluxDB extends InfluxDB {
  private __stackedPoints: IPoint[];
  constructor(options: any, {debounceWait = 1000}: BufferedInfluxDBOptions = {}) {
    super(options);
    this.__stackedPoints = [];
    this.__debouncedFlushPoints = debounce(this.__flushPoints, debounceWait, {maxWait: debounceWait});
  }
  async writePoints(points: IPoint[], options?: IWriteOptions): Promise<void> {
    this.__stackedPoints.push(...points);
    this.__debouncedFlushPoints(options);
  }
  async __flushPoints(options?: IWriteOptions): Promise<void> {
    const stackedPointsSoFar = this.__stackedPoints.splice(0, this.__stackedPoints.length);
    const uniqueStackedPoints = withUniqueTimestamps(stackedPointsSoFar);
    try {
      return await super.writePoints(uniqueStackedPoints, options);
    } catch (err) {
      // Restore stack on error
      this.__stackedPoints.unshift(...stackedPointsSoFar);
      throw err;
    }
  }
  // typings
  async __writePoints(_points: IPoint[], _options?: IWriteOptions): Promise<void> {} // eslint-disable-line
  async __debouncedFlushPoints(_options?: IWriteOptions): Promise<void> {} // eslint-disable-line
}

export default BufferedInfluxDB;
