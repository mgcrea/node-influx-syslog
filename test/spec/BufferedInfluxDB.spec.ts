import {FieldType, __mockedWritePoints} from 'influx';
import BufferedInfluxDB from '../../src/utils/BufferedInfluxDB';
import {waitFor} from '../utils';

const INFLUX_SCHEMA = {
  measurement: 'response_times',
  fields: {
    path: FieldType.STRING,
    duration: FieldType.INTEGER
  },
  tags: ['host']
};

describe('BufferedInfluxDB', () => {
  it('should properly export a class', () => {
    expect(BufferedInfluxDB).toBeDefined();
    expect(typeof BufferedInfluxDB).toBe('function');
  });
  describe('should properly connect', () => {
    const influx = new BufferedInfluxDB('http://localhost:8086/syslog', {debounceWait: 100});
    influx.addSchema(INFLUX_SCHEMA);
    beforeEach(() => {
      jest.resetAllMocks();
    });
    it('should only call parent writePoints after wait delay', async () => {
      await influx.writePoints([
        {
          measurement: 'response_times',
          tags: {host: 'foobar'},
          fields: {duration: 1000, path: '/foo/bar'}
        }
      ]);
      await waitFor(50);
      expect(influx['__stackedPoints'].length).toEqual(1);
      expect(influx['__stackedPoints']).toMatchSnapshot();
      expect(__mockedWritePoints).toHaveBeenCalledTimes(0);
      await waitFor(100);
      expect(influx['__stackedPoints'].length).toEqual(0);
      expect(influx['__stackedPoints']).toMatchSnapshot();
      expect(__mockedWritePoints).toHaveBeenCalledTimes(1);
    });
    it('should only call parent writePoints once after multiple writePoints', async () => {
      await influx.writePoints([
        {
          measurement: 'response_times',
          tags: {host: 'foobar'},
          fields: {duration: 1000, path: '/foo/bar'}
        }
      ]);
      await waitFor(10);
      await influx.writePoints([
        {
          measurement: 'response_times',
          tags: {host: 'foobar'},
          fields: {duration: 1000, path: '/foo/baz'}
        }
      ]);
      expect(influx['__stackedPoints'].length).toEqual(2);
      expect(influx['__stackedPoints']).toMatchSnapshot();
      expect(__mockedWritePoints).toHaveBeenCalledTimes(0);
      await waitFor(100);
      expect(influx['__stackedPoints'].length).toEqual(0);
      expect(influx['__stackedPoints']).toMatchSnapshot();
      expect(__mockedWritePoints).toHaveBeenCalledTimes(1);
    });
  });
});
