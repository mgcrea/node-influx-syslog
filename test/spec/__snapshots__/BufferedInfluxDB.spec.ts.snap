// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BufferedInfluxDB should properly connect should only call parent writePoints after wait delay 1`] = `
Array [
  Object {
    "fields": Object {
      "duration": 1000,
      "path": "/foo/bar",
    },
    "measurement": "response_times",
    "tags": Object {
      "host": "foobar",
    },
  },
]
`;

exports[`BufferedInfluxDB should properly connect should only call parent writePoints after wait delay 2`] = `Array []`;

exports[`BufferedInfluxDB should properly connect should only call parent writePoints once after multiple writePoints 1`] = `
Array [
  Object {
    "fields": Object {
      "duration": 1000,
      "path": "/foo/bar",
    },
    "measurement": "response_times",
    "tags": Object {
      "host": "foobar",
    },
  },
  Object {
    "fields": Object {
      "duration": 1000,
      "path": "/foo/baz",
    },
    "measurement": "response_times",
    "tags": Object {
      "host": "foobar",
    },
  },
]
`;

exports[`BufferedInfluxDB should properly connect should only call parent writePoints once after multiple writePoints 2`] = `Array []`;
